File: java/src/test/java/challenges/aoc2024/day3/model/ComputerMemoryTest.java
package challenges.aoc2024.day3.model;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;

class ComputerMemoryTest {
    
    @Test
    void shouldBuildComputerMemoryObjectFromString() {
        // When
        final ComputerMemory memory = ComputerMemory.of("xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))");

        // Then
        assertEquals(ImmutableComputerMemory.class, memory.getClass());

        assertEquals(4, memory.getInstructions().size());
        assertEquals(MultiplicationInstruction.of(2, 4), memory.getInstructions().get(0));
        assertEquals(MultiplicationInstruction.of(5, 5), memory.getInstructions().get(1));
        assertEquals(MultiplicationInstruction.of(11, 8), memory.getInstructions().get(2));
        assertEquals(MultiplicationInstruction.of(8, 5), memory.getInstructions().get(3));
    }
}

File: scala/src/test/scala/challenges/aoc2024/day3/model/ComputerMemoryTest.scala
package challenges.aoc2024.day3.model

import org.scalatest.funsuite.AnyFunSuite

class ComputerMemoryTest extends AnyFunSuite {

  test("should build ComputerMemorySection object from string") {
    // When
    val memory: ComputerMemorySection = ComputerMemorySection.of("xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))")

    // Then
    assert(memory.isInstanceOf[ComputerMemorySection])
    assert(4 == memory.instructions.size)
    assert(MultiplicationInstruction(2, 4) == memory.instructions.head)
    assert(MultiplicationInstruction(5, 5) == memory.instructions(1))
    assert(MultiplicationInstruction(11, 8) == memory.instructions(2))
    assert(MultiplicationInstruction(8, 5) == memory.instructions(3))
  }
}
